compiler/src/iree/compiler/GlobalOptimization/PropagateLinalgTranspose.cpp

```diff
@@ -113,7 +116,233 @@ static void specializeGenericTransposeOp(RewriterBase &rewriter,
 }
 
 //===----------------------------------------------------------------------===//
-// Transpose propagation
+// Other pattern helpers
+//===----------------------------------------------------------------------===//
+
+/// If the `op` is a ContractionOpInterface, return the generalized op if
+/// generalizing is allowed. Otherwise if the `op` is a linalg::GenericOp,
+/// then just return the generic op.
+static FailureOr<linalg::GenericOp>
+getGenericOpOrGeneralizeContraction(RewriterBase &rewriter, Operation *op,
+                                    bool allowGeneralizing) {
+  auto linalgOp = dyn_cast<linalg::LinalgOp>(op);
+  if (!linalgOp) {
+    return failure();
+  }
+  // TODO: Right now this is restricted to contractions due to fragility around
+  // handling of convolutions.
+  if (!isa<linalg::GenericOp>(linalgOp) &&
+      !(allowGeneralizing && linalg::isaContractionOpInterface(linalgOp))) {
+    return failure();
+  }
+  auto genericOp = dyn_cast<linalg::GenericOp>(op);
+  if (genericOp) {
+    return genericOp;
+  }
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(linalgOp);
+  return linalg::generalizeNamedOp(rewriter, linalgOp);
+}
+
+//===----------------------------------------------------------------------===//
+// Transpose Bubbling Patterns
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+// Fuses a transpose with the init of a linalg.generic op or contraction op.
+// Contraction ops are generalized and then treated as a generic. For example,
+//
+// linalg.generic {
+//   indexing_maps = [affine_map<(d0, d1, d2) -> (d1, d0, d2)>,
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x7x5>) outs(%1 : tensor<7x2x5>)
+//
+// %2 = linalg.transpose ... permutation = [0, 2, 1] :
+//                           tensor<7x2x5> -> tensor<7x5x2>
+// Becomes
+//
+// linalg.generic {
+//   indexing_maps = [affine_map<(d0, d1, d2) -> (d2, d0, d1)>,
+//                    affine_map<(d0, d1, d2) -> (d0, d1, d2)>]
+//   ins(%0 : tensor<2x7x5>) outs(%3 : tensor<7x5x2>)
+class BubbleTransposeThroughDpsInit
```

Comment:
Nit: This is not actually bubbling the transpose through. This is folding the transpose. I think this could just reuse the elementwise op fusion logic if you just generalize the transpose and fuse the operations.

I am going to assume all of this logic is correct, but would be good to not duplicate all the logic that should already be implemented upstream and reuse it (unless I am missing something).

Reply:
I tried this but the upstream logic only works for elementwise ops and this pattern also works for producers with reduction loops. This kind of fusion is actually unique to transposes (and other data-movement only ops like broadcast) because it is only changing the order in which data is written back in, so I'm not sure there will be much to reuse.